lnwallet/chancloser/rbf_coop_transitions.go

```diff
@@ -0,0 +1,1117 @@
+		if err != nil {
+			return nil, err
+		}
+
+		// With our signature created, we'll now attempt to finalize
+		// the close process.
+		//
+		// TODO(roasbef); duplication
+		closeTx, _, err := env.CloseSigner.CompleteCooperativeClose(
+			localSig, remoteSig,
+			l.CloseChannelTerms.LocalDeliveryScript,
+			l.CloseChannelTerms.RemoteDeliveryScript,
+			msg.SigMsg.FeeSatoshis, chanOpts...,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		chancloserLog.Infof("ChannelPoint(%v): received sig (fee=%v "+
+			"sats) from remote party, signing new tx=%v",
+			env.ChanPoint, msg.SigMsg.FeeSatoshis,
+			newLogClosure(func() string {
+				return spew.Sdump(closeTx)
+			}),
+		)
+
+		var closingSigs lnwire.ClosingSigs
+		if noClosee {
+			closingSigs.CloserNoClosee = newSigTlv[tlv.TlvType1](wireSig)
+		} else {
+			closingSigs.CloserAndClosee = newSigTlv[tlv.TlvType3](wireSig)
+		}
+
+		// As we're about to broadcast a new version of the co-op close
+		// transaction, we'll mark again as broadcast, but with this
+		// variant of the co-op close tx.
+		//
+		// TODO(roasbeef): db will only store one instance, store both?
+		err = env.ChanObserver.MarkCoopBroadcasted(closeTx, false)
+		if err != nil {
+			return nil, err
+		}
+
+		// As we transition, we'll omit two events: one to broadcast
+		// the transaction, and the other to send our ClosingSig
+		// message to the remote party.
+		sendEvent := &protofsm.SendMsgEvent[ProtocolEvent]{
+			TargetPeer: env.ChanPeer,
+			Msgs: []lnwire.Message{&lnwire.ClosingSig{
+				ChannelID:   env.ChanID,
+				ClosingSigs: closingSigs,
+			}},
+		}
+		broadcastEvent := &protofsm.BroadcastTxn{
+			Tx: closeTx,
+			Label: labels.MakeLabel(
+				labels.LabelTypeChannelClose, &env.Scid,
+			),
+		}
+		daemonEvents := protofsm.DaemonEventSet{sendEvent, broadcastEvent}
+
+		// Now that we've extracted the signature, we'll transition to
+		// the next state where we'll sign+broadcast the sig.
+		return &CloseStateTransition{
+			NextState: &ClosePending{
+				CloseTx: closeTx,
+			},
+			NewEvents: fn.Some(protofsm.EmittedEvent[ProtocolEvent]{
+				ExternalEvents: fn.Some(daemonEvents),
+			}),
+		}, nil
+	}
+
+	return nil, fmt.Errorf("%w: received %T while in RemoteCloseStart",
+		ErrInvalidStateTransition, event)
+}
+
+// ProcessEvent is a semi-terminal state in the rbf-coop close state machine.
+// In this state, we're waiting for either a confirmation, or for either side
+// to attempt to create a new RBF'd co-op close transaction.
+func (c *ClosePending) ProcessEvent(event ProtocolEvent, env *Environment,
+) (*CloseStateTransition, error) {
+
+	switch msg := event.(type) {
+	// If we can a spend while waiting for the close, then we'll go to our
+	// terminal state.
+	case *SpendEvent:
+		return &CloseStateTransition{
+			NextState: &CloseFin{
+				transitionEvent: msg,
+				ConfirmedTx:     msg.Tx,
+			},
+		}, nil
+
+	default:
+
+		return &CloseStateTransition{
+			NextState: c,
+		}, nil
+	}
```

Comment:
Why not error here like we do elsewhere to mark invalid state transitions? I think we should stay consistent about whether we silently drop the transition or we return an error. I think mixing approaches will lead to confusion.

Reply:
Further, this seems to imply that a new `Shutdown` here won't kick the process off again.

Reply:
Still relevant, seems weird that we process out-of-order events here as a noop but elsewhere as an error.

Reply:
So this is a terminal state for the forked off state (one local, one remote) we create once we start negotiation. See the `ClosingNegotiation` state. If that outer state gets a `Shutdown`, then we can restart as normal.